File: src/testmod/java/net/frozenblock/lib/testmod/config/TestConfig.java
/*
 * Copyright 2023-2024 FrozenBlock
 * This file is part of FrozenLib.
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, see <https://www.gnu.org/licenses/>.
 */

package net.frozenblock.lib.testmod.config;

import blue.endless.jankson.Comment;
import com.mojang.serialization.Codec;
import java.util.List;
import net.fabricmc.api.EnvType;
import net.fabricmc.api.Environment;
import net.frozenblock.lib.config.api.entry.TypedEntry;
import net.frozenblock.lib.config.api.entry.TypedEntryType;
import net.frozenblock.lib.config.api.instance.Config;
import net.frozenblock.lib.config.api.instance.json.JsonConfig;
import net.frozenblock.lib.config.api.instance.json.JsonType;
import net.frozenblock.lib.config.api.instance.toml.TomlConfig;
import net.frozenblock.lib.config.api.registry.ConfigRegistry;
import net.frozenblock.lib.testmod.FrozenTestMain;
import net.minecraft.client.gui.screens.Screen;
import net.minecraft.sounds.SoundEvent;
import net.minecraft.sounds.SoundEvents;
import net.minecraft.world.phys.Vec3;

public class TestConfig {

	// Make sure to define TypedEntryTypes at top of class.

	public static final Config<TestConfig> INSTANCE = ConfigRegistry.register(
			new TomlConfig<>(
				FrozenTestMain.MOD_ID,
				TestConfig.class
			)
	);

	@Comment("This is a boolean value.")
	public boolean testToggle = true;

	@Comment("This is an integer value.")
	public int testInt = 69;

	@Comment("This is a long value.")
	public long testLong = 69420L;

	@Comment("This is a float value.")
	public float testFloat = 69.420F;

	@Comment("This is a double value.")
	public double testDouble = 69.4206942069420D;

	@Comment("This is an integer list typed entry.")
	public List<Integer> testIntList = List.of(45);

	@Comment("This is a sound event typed entry.")
	public SoundEvent randomSound = SoundEvents.BEE_LOOP;

	@Comment("This is a Vec3 list typed entry.")
	public List<Vec3> typedVecList = List.of(new Vec3(0, 0, 0), new Vec3(1, 1, 1));

	@Comment("This is a list of doubles")
	public List<Double> doubleList = List.of(1D, 2D, 3D, Math.PI);

	@Comment("Sub menu!")
	public SubMenu subMenu = new SubMenu();

	public static class SubMenu {
		@Comment("Crazy sub option ngl")
		public boolean subOption = true;
	}

	public static TestConfig get() {
		return get(false);
	}

	public static TestConfig get(boolean real) {
		if (real) return INSTANCE.instance();
		return INSTANCE.config();
	}
}

File: src/testmod/java/net/frozenblock/lib/testmod/config/TestConfig.kt
/*
 * Copyright (C) 2024-2026 FrozenBlock
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <https://www.gnu.org/licenses/>.
 */

package net.frozenblock.lib.testmod.config

import blue.endless.jankson.Comment
import com.mojang.serialization.Codec
import com.mojang.serialization.codecs.RecordCodecBuilder
import io.netty.buffer.ByteBuf
import net.frozenblock.lib.FrozenLibConstants
import net.frozenblock.lib.config.v2.config.ConfigData
import net.frozenblock.lib.config.v2.config.ConfigSettings
import net.frozenblock.lib.config.v2.entry.ConfigEntry
import net.frozenblock.lib.config.v2.entry.EntryType
import net.frozenblock.lib.config.v2.entry.property.VisibilityPredicate
import net.minecraft.core.Holder
import net.minecraft.core.registries.BuiltInRegistries
import net.minecraft.core.registries.Registries
import net.minecraft.network.codec.ByteBufCodecs
import net.minecraft.network.codec.StreamCodec
import net.minecraft.sounds.SoundEvent
import net.minecraft.sounds.SoundEvents
import net.minecraft.world.phys.Vec3

private val SOUND_ENTRY_TYPE: EntryType<Holder<SoundEvent>> = EntryType.create(
    BuiltInRegistries.SOUND_EVENT.holderByNameCodec(), ByteBufCodecs.holderRegistry(Registries.SOUND_EVENT)
)

object TestConfig {
    @JvmField
    val CONFIG: ConfigData<*> = ConfigData.createAndRegister(
        FrozenLibConstants.config("testmod"),
        ConfigSettings.DJS
    )
    @JvmField
    @Comment("This is a boolean value.")
    var testToggle: ConfigEntry<Boolean> = CONFIG.entry("testToggle", EntryType.BOOL, true)

    @JvmField
    @Comment("This is an integer value.")
    var testInt: ConfigEntry<Int> = CONFIG.entry("testInt", EntryType.INT, 69)

    @JvmField
    @Comment("This is a long value.")
    var testLong: ConfigEntry<Long> = CONFIG.entry("testLong", EntryType.LONG, 69420L)

    @JvmField
    @Comment("This is a float value.")
    var testFloat: ConfigEntry<Float> = CONFIG.entry("testFloat", EntryType.FLOAT, 69.420f)

    @JvmField
    @Comment("This is a double value.")
    var testDouble: ConfigEntry<Double> = CONFIG.entry("testDouble", EntryType.DOUBLE, 69.4206942069420)

    @JvmField
    @Comment("This is an integer list typed entry.")
    var testIntList: ConfigEntry<List<Int>> = CONFIG.entry("testIntList", EntryType.INT.asList(), listOf(45))

    @JvmField
    @Comment("This is a sound event typed entry.")
    var randomSound: ConfigEntry<Holder<SoundEvent>> = CONFIG.entry(
        "randomSound",
        SOUND_ENTRY_TYPE,
        SoundEvents.BREEZE_WIND_CHARGE_BURST
    )

    @JvmField
    @Comment("This is a Vec3 list typed entry.")
    var typedVecList: ConfigEntry<List<Vec3>> = CONFIG.entry(
        "typedVecList",
        EntryType.VEC3.asList(),
        listOf(
            Vec3(0.0, 0.0, 0.0),
            Vec3(1.0, 1.0, 1.0)
        )
    )

    @JvmField
    @Comment("This is a list of doubles")
    var doubleList: ConfigEntry<List<Double>> = CONFIG.entry("doubleList", EntryType.DOUBLE.asList(), listOf(1.0, 2.0, 3.0, Math.PI))

    @JvmField
    @Comment("This should not be seen unless config serialization is bugged")
    var mysteriousUnknownToggle: ConfigEntry<Boolean> = CONFIG.entryBuilder("mysteriousUnknownToggle", EntryType.BOOL, false)
        .visibilityPredicate(VisibilityPredicate.of { false })
        .build()

    @JvmField
    @Comment("Sub menu!")
    var subMenu: ConfigEntry<SubMenu> = CONFIG.entry("subMenu", SubMenu.ENTRY_TYPE, SubMenu())

    data class SubMenu(
        @JvmField
        @Comment("Crazy sub option ngl")
        var subOption: Boolean = true
    ) {
        companion object {
            val CODEC: Codec<SubMenu> = RecordCodecBuilder.create { instance ->
                instance.group(
                    Codec.BOOL.fieldOf("subOption").forGetter { it.subOption }
                ).apply(instance, ::SubMenu)
            }
            val STREAM_CODEC: StreamCodec<ByteBuf, SubMenu> = StreamCodec.composite(
                ByteBufCodecs.BOOL, SubMenu::subOption, ::SubMenu
            )
            val ENTRY_TYPE: EntryType<SubMenu> = EntryType.create(CODEC, STREAM_CODEC)
        }
    }
}
